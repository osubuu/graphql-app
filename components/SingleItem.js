import { Query } from 'react-apollo';
import gql from 'graphql-tag';
import styled from 'styled-components';
import Head from 'next/head';
import Link from 'next/link';
import { adopt } from 'react-adopt';
import User from './User';

import ErrorMessage from './ErrorMessage';
import formatMoney from '../lib/formatMoney';

import DeleteItem from './DeleteItem';
import AddToCart from './AddToCart';

const SingleItemStyles = styled.div`
  max-width: 1200px;
  margin: 2rem auto;
  box-shadow: ${props => props.theme.bs};
  display: flex;
  flex-direction: column;
  align-items: center;
  min-height: 800px;

  .image-container{
    width: 30rem;
    margin-top: 1rem;

    img {
      width: 100%;
      height: 100%;
      object-fit: contain;
    }
  }

  .details {
    margin: 3rem;
    font-size: 2rem;
    text-align: center;
  }

  .buttonList {
    display: grid;
    width: 80%;
    border: 1px solid ${props => props.theme.lightGrey};
    grid-template-columns: repeat(auto-fit, minmax(100px, 1fr));
    grid-gap: 1px;
    background: ${props => props.theme.lightGrey};
    margin-top: 4rem;
    & > * {
      text-align: center;
      background: white;
      border: 0;
      font-size: 1rem;
      padding: 1rem;
      cursor: pointer;

      &:hover {
        color: ${props => props.theme.teal};
      }
    }
  }
`;

export const SINGLE_ITEM_QUERY = gql`
  query SINGLE_ITEM_QUERY($id: ID!) {
    item(where: {id: $id}) {
      id
      title
      description
      largeImage
      price
      user {
        id
      }
    }
  }
`;

const SingleItem = props => {
  const Composed = adopt({
    user: ({ render }) => <User>{render}</User>,
    singleItem: ({ render }) => (
      <Query
        query={SINGLE_ITEM_QUERY}
        variables={{ id: props.id }}
      >{render}
      </Query>
    ),
  });

  return (
    <Composed>
      {({ user, singleItem }) => {
        if (singleItem.error) return <ErrorMessage error={singleItem.error} />;
        if (singleItem.loading) return <p>Loading...</p>;
        if (!singleItem.data.item) return <p>No Item Found for {props.id}</p>;
        const { item } = singleItem.data;
        const canEditAndDelete = user.data.me && user.data.me.id === item.user.id;
        return (
          <SingleItemStyles>
            <Head>
              <title>Hoodify - {item.title}</title>
            </Head>
            <div className="details">
              <h2>Viewing: {item.title}</h2>
              <h3>{formatMoney(item.price)}</h3>
              <p>{item.description}</p>
            </div>
            <div className="image-container">
              <img src={item.largeImage} alt={item.title} />
            </div>
            <div className="buttonList">
              {canEditAndDelete && (
              <Link href={{
                pathname: 'update',
                query: { id: item.id },
              }}
              >
                <a>Edit</a>
              </Link>
              )}
              <AddToCart id={item.id} />
              {canEditAndDelete && <DeleteItem id={item.id}>Delete This Item</DeleteItem>}
            </div>
          </SingleItemStyles>
        );
      }}
    </Composed>
  );
};

export default SingleItem;
