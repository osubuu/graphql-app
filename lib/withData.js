import withApollo from 'next-with-apollo';
import ApolloClient from 'apollo-boost';
import { endpoint, prodEndpoint } from '../config';
import { LOCAL_STATE_QUERY } from '../components/Cart';

function createClient({ headers }) {
  return new ApolloClient({
    uri: process.env.NODE_ENV === 'development' ? endpoint : prodEndpoint,
    request: operation => {
      operation.setContext({
        fetchOptions: {
          credentials: 'include',
        },
        headers,
      });
    },
    // local data
    clientState: {
      resolvers: {
        Mutation: {
          toggleCart(_, variables, client) { // _ is just a variable we don't use
            // read the cartOpen value from the cache
            const { cartOpen } = client.cache.readQuery({
              query: LOCAL_STATE_QUERY,
            });
            // write the cart state to the opposite
            const data = {
              data: { cartOpen: !cartOpen },
            };
            client.cache.writeData(data);
            return data;
          },
        },
      },
      defaults: {
        cartOpen: false,
      },
    },
  });
}

export default withApollo(createClient);
