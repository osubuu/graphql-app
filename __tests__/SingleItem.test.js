import { mount } from 'enzyme';
import toJSON from 'enzyme-to-json';
import wait from 'waait';
import { MockedProvider } from 'react-apollo/test-utils';

import SingleItem, { SINGLE_ITEM_QUERY } from '../components/SingleItem';
import { fakeItem } from '../lib/testUtils';

describe('<SingleItem/>', () => {
  it('renders with proper data', async () => {
    const mocks = [
      {
        // when someone makes a request with this query and variable combo
        request: { query: SINGLE_ITEM_QUERY, variables: { id: '123' } },
        // return this fake data (mocked data)
        result: {
          data: {
            item: fakeItem(),
          },
        },
      },
    ];
    const wrapper = mount(
      <MockedProvider mocks={mocks}>
        <SingleItem id="123" />
      </MockedProvider>,
    );
    expect(wrapper.text()).toContain('Loading...');

    // package to wait x amount of milliseconds
    // to ensure the following code is put at the end of the JS event callstack
    await wait();

    // manually update the wrapper
    wrapper.update();

    expect(toJSON(wrapper.find('h2'))).toMatchSnapshot();
    expect(toJSON(wrapper.find('img'))).toMatchSnapshot();
    expect(toJSON(wrapper.find('p'))).toMatchSnapshot();
  });

  it('errors with a not found item', async () => {
    const mocks = [
      {
        request: { query: SINGLE_ITEM_QUERY, variables: { id: '123' } },
        result: {
          errors: [{ message: 'Items not found!' }],
        },
      },
    ];

    const wrapper = mount(
      <MockedProvider mocks={mocks}>
        <SingleItem id="123" />
      </MockedProvider>,
    );

    await wait();
    wrapper.update();
    const item = wrapper.find('[data-test="graphql-error"]');
    expect(item.text()).toContain('Items not found!');
    expect(toJSON(item)).toMatchSnapshot();
  });
});
