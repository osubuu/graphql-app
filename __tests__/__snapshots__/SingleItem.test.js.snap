// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SingleItem/> errors with a not found item 1`] = `
<p
  data-test="graphql-error"
>
  <strong>
    Shoot!
  </strong>
  Items not found!
</p>
`;

exports[`<SingleItem/> renders with proper data 1`] = `null`;

exports[`<SingleItem/> renders with proper data 2`] = `null`;

exports[`<SingleItem/> renders with proper data 3`] = `null`;
